File: mobile/android/app/src/main/java/app/terra/explore/LiveSpeed.kt
package app.terra.explore
class LiveSpeed {
    private val samples=mutableListOf<Pair<Long,Double>>()
    fun add(metersPerSecond: Double,accuracy: Double,time: Long,maximum: Double) {
        if(!metersPerSecond.isFinite() || metersPerSecond<0 || metersPerSecond>maximum || accuracy !in 0.0..3.0 || time<=(samples.lastOrNull()?.first ?: 0)) return
        samples.removeAll { time-it.first>3000 }; samples.add(time to metersPerSecond*3.6)
        if(metersPerSecond<0.3) { samples.clear(); samples.add(time to 0.0) }
        if(samples.size>3) samples.removeAt(0)
    }
    fun value(now: Long): Double? { val last=samples.lastOrNull() ?: return null; if(now-last.first !in -1000..4000) return null; return samples.map { it.second }.sorted()[samples.size/2] }
    fun reset() { samples.clear() }
}

File: mobile/ios/Terra/LiveSpeed.swift
import Foundation
struct LiveSpeed {
    private var samples: [(Double,Double)] = []
    mutating func add(metersPerSecond: Double, accuracy: Double, time: Double, maximum: Double) {
        guard metersPerSecond.isFinite, metersPerSecond >= 0, metersPerSecond <= maximum, accuracy >= 0, accuracy <= 3, time > (samples.last?.0 ?? 0) else { return }
        samples.removeAll { time-$0.0 > 3000 }; samples.append((time,metersPerSecond*3.6))
        if metersPerSecond < 0.3 { samples = [(time,0)] }
        if samples.count > 3 { samples.removeFirst() }
    }
    func value(now: Double) -> Double? {
        guard let last = samples.last, now-last.0 >= -1000, now-last.0 <= 4000 else { return nil }
        let sorted = samples.map { $0.1 }.sorted(); return sorted[sorted.count/2]
    }
    mutating func reset() { samples.removeAll() }
}
